analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs

```diff
@@ -0,0 +1,62 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ArrayPassedAsParams : ArrayPassedAsParamsBase<SyntaxKind, ArgumentSyntax>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    protected override SyntaxKind[] ExpressionKinds { get; } =
+        {
+            SyntaxKind.ObjectCreationExpression,
+            SyntaxKind.InvocationExpression,
+            SyntaxKindEx.ImplicitObjectCreationExpression
+        };
+
+    protected override ArgumentSyntax GetLastArgumentIfArrayCreation(SyntaxNode expression) =>
+        GetLastArgumentIfArrayCreation(GetArgumentListFromExpression(expression));
+
+    private static BaseArgumentListSyntax GetArgumentListFromExpression(SyntaxNode expression) =>
+        expression switch
+        {
+            ObjectCreationExpressionSyntax { } creation => creation.ArgumentList,
+            InvocationExpressionSyntax { } invocation => invocation.ArgumentList,
+            _ when ImplicitObjectCreationExpressionSyntaxWrapper.IsInstance(expression) =>
+                ((ImplicitObjectCreationExpressionSyntaxWrapper)expression).ArgumentList,
+            _ => null
+        };
+
+    private static ArgumentSyntax GetLastArgumentIfArrayCreation(BaseArgumentListSyntax argumentList) =>
+        argumentList is { Arguments: { Count: > 0 } arguments }
+        && IsArrayCreation(arguments.Last().Expression)
+            ? arguments.Last()
+            : null;
+
+    private static bool IsArrayCreation(ExpressionSyntax expression) =>
+        expression switch
+        {
+            ArrayCreationExpressionSyntax { } arrayCreation =>
+                arrayCreation.Initializer is InitializerExpressionSyntax { Expressions.Count: > 0 },
```

Comment:
What do you think about removing the `Expressions.Count: > 0` restriction for C#? That way we would complain about `new int[] { }` also. I don't see any problem, except maybe for overload resolution related disambiguations like demonstrated [here](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgZgAIMiBhIgbzyNpPXRrutzraIFkAKTASgG4m7TjxhEoAUwDuRAJZRgAbQC6VIgF8BRAPTaiAMQD2YAMYSiwABbmADgEMwdgLYBnIoYBuEsABtDduAtDIgAjcxM7Hx8JOCENOLjUTAA2EgAWEXlgOV4qOLYkgE4uACIAOWD7R1d3L19/OBKBOPU2PESU9MyFOTEq5zcslSIbXJZhWiLSgAUHAdrvPwCmwVY6dTx1IA).

Reply:
I don't have a strong opinion on that, I'm not able to estimate how frequent situations like this overload resolution are. I have the feeling that the vast majority are indeed TPs and we probably should report. At this point, I'll go ahead and change this behavior, if we'll receive some FPs complaints from the community it should be an easy fix to revert it to this state.

Reply:
👍 There were no new issues coming up in the ITs after the change. This indicates that it is okay to do it. We need to do peach validation anyway and if we see any problems there, we can revert it right away. I think we have a nice symmetry between C# and VB now.